### PR TITLE
Compute layout for initially undisplayed components

### LINKED ISSUE #12640: [BUG] Embedded plot not sized correctly if not visible from the start
### Software versions

Python version      :  3.8.5 (default, Sep  4 2020, 07:30:14) 
IPython version     :  8.6.0
Tornado version     :  6.2
Bokeh version       :  3.0.2
BokehJS static path :  /home/michael/work/imtek/issues/bokeh_3.0.1_problem/venv/lib/python3.8/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  Linux-5.4.0-132-generic-x86_64-with-glibc2.10

### Browser name and version

Firefox 107.0

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

In my web app, I want to show two bokeh plots, each in a separate bootstrap tab. The are embedded
using [*components*](https://docs.bokeh.org/en/latest/docs/user_guide/output/embed.html#components).

When changing the tab, the corresponding plot in the tab should always be rendered and visible with correct size,
no matter whether it was visible from the start or not.

This worked fine with bokeh version 2.4.3, but not anymore with bokeh 3.0.x.

### Observed behavior

The plot on the second tab (which was not active when loading the page) shows a degenerated plot only.
The plot on the first tab looks good.
See series of screenshots below.

### Example code

```Python
############################################################
## Preparation
############################################################
pip install bokeh==3.0.2
pip install flask

############################################################
## File "app.py"
############################################################

from flask import Flask, render_template
from bokeh.embed import components
from bokeh.plotting import figure
from bokeh.resources import INLINE

app = Flask(__name__)

@app.route('/')
def index():

    fig1 = figure(width=300, height=300)
    fig1.line(
        x=[1, 2, 3, 4],
        y=[4, 3, 5, 1],
    )

    fig2 = figure(width=300, height=300)
    fig2.line(
        x=[1, 2, 3, 4],
        y=[1, -3, 5, -1],
    )

    
    # grab the static resources
    js_resources = INLINE.render_js()
    css_resources = INLINE.render_css()

    # render template
    script1, div1 = components(fig1)
    script2, div2 = components(fig2)
    html = render_template(
        'index.html',
        plot_script1=script1,
        plot_div1=div1,
        plot_script2=script2,
        plot_div2=div2,
        js_resources=js_resources,
        css_resources=css_resources,
    )
    return html
    

if __name__ == '__main__':
    app.run(port=5001, debug=True)

############################################################
## File "templates/index.html" (uses bootstrap via CDN)
############################################################
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="content-type" content="text/html; charset=utf-8">
    <title>Bokeh Tab Issue</title>

    <!-- Bokeh resources from flask -->
    {{ js_resources|indent(4)|safe }}
    {{ css_resources|indent(4)|safe }}
    {{ plot_script1|indent(4)|safe }}
    {{ plot_script2|indent(4)|safe }}

    <!-- Bootstrap 4 for tab navigation -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/css/bootstrap.min.css" crossorigin="anonymous">
    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.12.9/dist/umd/popper.min.js" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/js/bootstrap.min.js" crossorigin="anonymous"></script>
  </head>
  <body>

    <nav class="nav nav-tabs">
      <a class="nav-item nav-link active" data-toggle="tab" href="#first">First</a>
      <a class="nav-item nav-link" data-toggle="tab" href="#second">Second</a>
    </nav>

    <div class="tab-content">
      <div class="tab-pane fade show active" id="first" role="tabpanel">
	<h2>On first tab</h2>
	{{ plot_div1|indent(4)|safe }}
      </div>
      <div class="tab-pane fade" id="second" role="tabpanel">
	<h2>On second tab</h2>
	{{ plot_div2|indent(4)|safe }}
      </div>
    </div>
       
  </body>
</html>
```

### Stack traceback or browser console output

No error message in Python or in JS console (except the favicon is missing).

## This is how the first tab looks like, which is fine:
grafik

## Using _bokeh 3.0.2_, the plot on the second tab is broken:
grafik

## Using _bokeh 2.4.3_ with exactly the same code, the second plot is also okay:
grafik

Note that the second plot has different data than the first in order to be sure it's the other figure.

### PR DESCRIPTION
**Pull request by mattpap**
fixes #12640

### CODE REVIEW

File: bokehjs/test/unit/models/canvas/canvas.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ describe("Canvas", () => {
   describe("should support composing layers", () => {
     it.dpr(1)("with devicePixelRatio == 1", async () => {
       const canvas = new Canvas({output_backend: "canvas", hidpi: true, styles: {width: "600px", height: "600px"}})
-      const {view: canvas_view} = await display(canvas, null)
+      const {view: canvas_view} = await display(canvas, [650, 650])
```
Comment (bryevdv) [inline bokehjs/test/unit/models/canvas/canvas.ts:11]: What is the rationale/need for these test changes?

**Comment by bryevdv** [comment]
Confirm this change fixes the MRE in #12640

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Assuming you don't want to make the *U00X* suggested updates in this PR, those rules can be disabled here:

https://github.com/bokeh/bokeh/blob/829a9d09c327164153a24e17c3d5a1d56c26f755/pyproject.toml#L117-L122

(Either remove "U" altogether or ignore specific rules)

Edit: if you *do* want to apply them, `ruff --fix .` will fix all but two issues.

File: bokehjs/test/unit/models/canvas/canvas.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/test/unit/models/canvas/canvas.ts:11]: `null` should be used when overflow is inevitable, in which case `null` disables overflow checks. This is not the case here, so there should be a viewport specified or left out for inferred viewport. Normally one wouldn't specify this altogether, but bare UI element don't allow viewport inference.

**Comment by mattpap** [comment]
Given that `pyproject.toml` contains `target-version = "py38"` for ruff, then most of those errors don't make sense. This must be some kind of regression in ruff. I will pin it the the last working version, which we should have done anyway, because we shouldn't be updating code quality in random PRs.

**Comment by mattpap** [comment]
For example this:
```py
-TypeOrInst: TypeAlias = Union[Type[T], T]
+TypeOrInst: TypeAlias = type[T] | T
 
-Init: TypeAlias = Union[T, UndefinedType, IntrinsicType]
+Init: TypeAlias = T | UndefinedType | IntrinsicType
```
will fail in 3.8, 3.9.

**Comment by bryevdv** [comment]
> I will pin it the the last working version, which we should have done anyway, because we shouldn't be updating code quality in random PRs.

👍 This seems advised. FWIW the `ruff` dev is very responsive, if you want to submit an issue it will likely be addressed quickly.

**Comment by bryevdv** [comment]
@mattpap do you know if this is the fix for the Dask issue? As soon as that is in, I plan to make a 3.0.4 release for them

**Comment by mattpap** [comment]
It may help in some cases, but not the reported ones.

**Pull request merged**